sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -77,16 +100,30 @@ template <typename Compare = std::less<>> class default_sorter {
   template <typename Group, typename T>
   T operator()([[maybe_unused]] Group g, T val) {
 #ifdef __SYCL_DEVICE_ONLY__
+    // Adjust the scratch pointer based on alignment of the type T.
     // Per extension specification if scratch size is less than the value
     // returned by memory_required then behavior is undefined, so we don't check
     // that the scratch size statisfies the requirement.
-    auto range_size = g.get_local_range().size();
-    size_t local_id = g.get_local_linear_id();
-    T *temp = reinterpret_cast<T *>(scratch.data());
-    ::new (temp + local_id) T(val);
-    sycl::detail::merge_sort(g, temp, range_size, comp,
-                             scratch.data() + range_size * sizeof(T));
-    val = temp[local_id];
+    void *scratch_ptr = scratch.data();
+    size_t space = scratch.size();
+    const size_t range_size = g.get_local_range().size();
+    const size_t local_id = g.get_local_linear_id();
+    scratch_ptr =
+        std::align(alignof(T), /* output storage and temporary storage */ 2 *
+                                   range_size * sizeof(T),
+                   scratch_ptr, space);
+    std::byte *aligned_scratch_ptr = static_cast<std::byte *>(scratch_ptr);
+    // Each work item processes two locations: in the output buffer and in the
+    // temp buffer.
+    T *local_copy = ::new (aligned_scratch_ptr + local_id * sizeof(T)) T(val);
```

Comment:
since `local_id * sizeof(T)` used twice (here and 1 line below) it can be useful to create an additional variable. It would also improve readability